src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt

```diff
@@ -102,6 +116,151 @@ class MigrateVisitService(
+    return sessionTemplates
+  }
+
+  private fun getProximityMinutes(sessionStartTime: LocalTime, startTime: LocalTime, sessionEndTime: LocalTime, endTime: LocalTime): Int {
+    return (
+      (sessionStartTime.toSecondOfDay() - startTime.toSecondOfDay()).absoluteValue +
+        (sessionEndTime.toSecondOfDay() - endTime.toSecondOfDay()).absoluteValue
+      ) / 60
+  }
+
+  private fun getSessionCategory(
+    migrateVisitRequest: MigrateVisitRequestDto,
+  ): String? {
+    with(migrateVisitRequest) {
+      val isInTheFuture = !startTimestamp.isBefore(LocalDateTime.now())
+      if (isInTheFuture) {
+        val sessionTemplatesByProximityOrder = getSessionTemplatesInTimeProximityOrder(
+          prisonCode,
+          startTimestamp.toLocalDate(),
+          startTimestamp.toLocalTime(),
+          endTimestamp.toLocalTime(),
+          visitRoom,
+        )
+
+        val sessionsWithCapacityGroups = sessionTemplatesByProximityOrder.filter { it.capacityGroup != null }
+        if (sessionsWithCapacityGroups.isNotEmpty()) {
+          return getNearestTemplateThatMatchesPrisonerLocation(migrateVisitRequest, sessionsWithCapacityGroups).capacityGroup
+        }
+      }
+    }
+    return null
+  }
+
+  fun getNearestTemplateThatMatchesPrisonerLocation(
+    migrateVisitRequest: MigrateVisitRequestDto,
+    sessionTemplates: List<SessionTemplate>,
+  ): SessionTemplate {
+    val sessionLocationTemplates = sessionService.filterSessionsTemplatesForLocation(sessionTemplates, migrateVisitRequest.prisonerId, migrateVisitRequest.prisonCode, true)
+    if (sessionLocationTemplates.isNotEmpty()) {
+      val templateLocationMap = createTemplateLocationMap(sessionLocationTemplates)
+
+      val countLevelsThatMatch: (PermittedSessionLocation) -> Int = { location: PermittedSessionLocation ->
+        with(location) {
+          levelFourCode?.let { 4 } ?: levelThreeCode?.let { 3 } ?: levelTwoCode?.let { 2 } ?: 1
+        }
+      }
+
+      val getBestMatch = Comparator<PermittedSessionLocation> { location1, location2 ->
+        countLevelsThatMatch(location2).compareTo(countLevelsThatMatch(location1))
+      }
+
+      val proximityComparator = Comparator<Pair<String, MutableList<PermittedSessionLocation>>> { paire1, paire2 ->
+        val sorted1 = paire1.second.sortedWith(getBestMatch)
+        val sorted2 = paire2.second.sortedWith(getBestMatch)
+        getBestMatch.compare(sorted1.first(), sorted2.first())
+      }
+
+      val sortedByLevelMatch = templateLocationMap.toList().sortedWith(proximityComparator)
+      return sessionLocationTemplates.find { it.reference == sortedByLevelMatch.first().first }!!
+    }
+    return sessionTemplates.first()
+  }
+
+  private fun createTemplateLocationMap(sessionLocationTemplates: List<SessionTemplate>): MutableMap<String, MutableList<PermittedSessionLocation>> {
+    val templateLocationMap = mutableMapOf<String, MutableList<PermittedSessionLocation>>()
+    sessionLocationTemplates.forEach { st ->
```

Comment:
`st` and `lg` should be expanded, picky I know 😄

Reply:
I don't think it's picky at all, I hate using st or lg,.. it's not self descriptive, will change..